admin/services/update/update.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package update implements update API.
+package update
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/percona/pmm/admin/cli/flags"
+	dockerCmd "github.com/percona/pmm/admin/commands/pmm/server/docker"
+	"github.com/percona/pmm/admin/pkg/docker"
+	"github.com/percona/pmm/api/updatepb"
+)
+
+// Make sure this does not include any directories, just a basename.
+const logsFileNamePattern = "upgrade.*.log"
+
+type Server struct {
+	// Context coming from cli commands. When cancelled, the command has been cancelled.
+	cliCtx             context.Context
+	dockerFn           dockerFunctions
+	dockerImage        string
+	gRPCMessageMaxSize uint32
+	updateInProgress   map[string]struct{}
+	updateMu           sync.RWMutex
+
+	updatepb.UnimplementedUpdateServer
+}
+
+// New returns new instance of Server.
+func New(ctx context.Context, dockerImage string, gRPCMessageMaxSize uint32) (*Server, error) {
+	d, err := docker.New(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Server{
+		cliCtx:             ctx,
+		dockerFn:           d,
+		dockerImage:        dockerImage,
+		gRPCMessageMaxSize: gRPCMessageMaxSize,
+		updateInProgress:   make(map[string]struct{}, 1),
+	}, nil
+}
+
+// StartUpdate starts PMM Server update.
+func (s *Server) StartUpdate(ctx context.Context, req *updatepb.StartUpdateRequest) (*updatepb.StartUpdateResponse, error) {
+	containerID := req.Hostname
+
+	logrus.Debugf("Inspecting container %s", containerID)
+	container, err := s.dockerFn.ContainerInspect(ctx, containerID)
+	if err != nil {
+		logrus.Errorf("Could not inspect container %s. Error: %v", containerID, err)
+		return nil, err
+	}
+
+	if !container.State.Running {
+		return nil, fmt.Errorf("container %s is not running", containerID)
+	}
```

Comment:
I think this is a good check.
You can be in a situation where you have multiple PMM Server containers in the system since you kept them after an upgrade, but they're stopped.
In these cases you may have accidentally specified the wrong container name. This error notifies you that you may have made a mistake.

In cases where you have just a single container and it's stopped for some reason, the command fails. A workaround is very simple though - you start the container and run the command again.

Reply:
but if you want to safely stop container and then update you won't be able to do that. Can we add a prompt instead of showing an error?

Reply:
Adding a prompt would not work in this case. The `pmm-server-upgrade` does not accept any stdin. It's basically a daemon.
We could add it to the cli though, if it's not already there.

Reply:
Got it, now it makes sense.
I think we should split PMM CLI and `pmm-server-upgrade` to make the code more obvious.
And we should extract exact upgrade logic to separate struct to be able to reuse it in PMM CLI and `pmm-server-upgrade`

Reply:
I believe that's what we're doing.
`pmm-server-upgrade` uses one command from PMM CLI which is a separate structure mapped as a command in cli.
https://github.com/percona/pmm/blob/2b18b80ee7288fabf2b6a8a9264a9906d91bcc32/admin/commands/pmm/server/docker/upgrade.go#L38-L48

It's initialized here
https://github.com/percona/pmm/blob/2b18b80ee7288fabf2b6a8a9264a9906d91bcc32/admin/services/update/update.go#L102-L105

Do you have in mind you recommend separating it even further so that it does not live in the command but is rather a completely separate structure?

Happy to split it but I believe it's going to add additional complexity without reducing maintenance complexity.

Reply:
it looks not obvious, to make it clear `dockerCmd` should be renamed to `dockerUpgradeCmd`

Reply:
The `dockerCmd` is related to all server docker commands such as install, upgrade, etc.
I see your point though. I renamed `New` to `NewUpgradeCommand` - makes sense?